security/pkg/nodeagent/sds/sdsservice.go

```diff
@@ -599,7 +600,8 @@ func sdsDiscoveryResponse(s *model.SecretItem, conID, resourceName string) (*xds
 
 func newSDSConnection(stream discoveryStream) *sdsConnection {
 	return &sdsConnection{
-		pushChannel: make(chan *sdsEvent, 1),
+		// Make the channel size to be > 1 to avoid deadlock
```

Comment:
this is very suspicious. What deadlock? This isn't fixing the problem it is just hiding it.

Pilot has the same push channel with a length of 1, if there is a deadlock we should fix it properly

Reply:
sg, can you explain events that trigger deadlock ?

Reply:
In the function StreamSecrets, the same thread is watching both reqChannel and -con.pushChannel. If the code responding to reqChannel push events to con.PushChannel, we got a deadlock because the thread would not take the event, and the function would not return due to the channel behavior. So in general, it is safer to make the channel size being greater than 1. I made it larger just so that we have a safety margin, and reduce the chance for multiple threads to wait for the channel as synchronization is not so important here (I think). This deadlock condition is triggered with current patch when we add secret in responding to reqChannel.

Reply:
thanks for the explanation. it doesn't seem like the right fix to me, but I'm saying that without having looked at the code. I've been working a lot on the same codepath in pilot so from my experience it's pretty delicate for deadlocks so want to be safe. Let me look into it a bit more

Reply:
If I don't add the secret to the cache, I don't need this change. I actually also tried to just return with the fetched secret without putting it into cache, and it worked too. The only caveat is that it is not so logical when we know that our cache missed a secret and we don't fix it. Also because it is not put in cache, it would cause multiple API call because the function would be called twice by ShouldWaitForIngressGatewaySecret() and the following GenerateSecret() - this can be fixed though.  If that solution is more preferred, I can provide a patch in that direction.

Reply:
We need to fix this in a way that does not involve increasing the channel size. This is not safe. What happens if 16 proxies that all push in the reqChannel code? I doubt that is even that rare and now the deadlock comes back

Reply:
The reqChannel is created per stream, and the selection function should make sure that no watched channel gets starved, so I don't know why it is an issue if 16 proxies all push in the reqChannel. Nevertheless, I submitted an updated patch so that we don't write the secret to cache and thus don't need to change the channel size. This patch is handling an unexpected/rare situation anyway, and the cache would be re-synced after the watcher recovers.